oak_functions/loader/src/server.rs

```diff
@@ -14,65 +14,207 @@
+                    err
+                );
+                OakStatus::ErrInvalidArgs
+            })?;
+
+        if dest_capacity >= self.request_bytes.len() as u32 {
+            self.get_memory().set(dest, &self.request_bytes).map_err(|err| {
+                    error!(
+                        "read_request(): Unable to write destination buffer into guest memory: {:?}",err
+                    );
+                    OakStatus::ErrInvalidArgs
+                })?;
+        } else {
+            return Err(OakStatus::ErrBufferTooSmall);
+        }
+
+        Ok(())
+    }
+
+    /// Corresponds to the host ABI function [`write_response`](https://github.com/project-oak/oak/blob/main/docs/oak_functions_abi.md#write_response).
+    pub fn write_response(
+        &mut self,
+        source: AbiPointer,
+        source_length: AbiPointerOffset,
+    ) -> Result<(), OakStatus> {
+        let response = self
+            .get_memory()
+            .get(source, source_length as usize)
+            .map_err(|err| {
+                error!(
+                    "write_response(): Unable to read name from guest memory: {:?}",
+                    err
+                );
+                OakStatus::ErrInvalidArgs
+            })?;
+        self.response_bytes = response;
+        Ok(())
+    }
 }
 
+impl wasmi::Externals for WasmInterface {
+    /// Invocation of a host function specified by its registered index. Acts as a wrapper for
+    /// the relevant native function, just:
+    /// - checking argument types (which should be correct as `wasmi` will only pass through those
+    ///   types that were specified when the host function was registered with `resolv_func`).
+    /// - mapping resulting return/error values.
+    fn invoke_index(
+        &mut self,
+        index: usize,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Option<wasmi::RuntimeValue>, wasmi::Trap> {
+        match index {
+            READ_REQUEST => map_host_errors(self.read_request(
+                args.nth_checked(0)?,
+                args.nth_checked(1)?,
+                args.nth_checked(2)?,
+            )),
+            WRITE_RESPONSE => {
+                map_host_errors(self.write_response(args.nth_checked(0)?, args.nth_checked(1)?))
+            }
+            _ => panic!("Unimplemented function at {}", index),
+        }
+    }
+}
+
+impl wasmi::ModuleImportResolver for WasmInterface {
+    fn resolve_func(
+        &self,
+        field_name: &str,
+        signature: &wasmi::Signature,
+    ) -> Result<wasmi::FuncRef, wasmi::Error> {
+        oak_resolve_func(field_name, signature)
+    }
+}
 /// Encapsulates the state of a Wasm invocation for a single user request.
 struct WasmState {
     instance: wasmi::ModuleRef,
-    #[allow(dead_code)]
-    memory: wasmi::MemoryRef,
-    #[allow(dead_code)]
-    request_bytes: Vec<u8>,
-    response_bytes: Vec<u8>,
+    abi: WasmInterface,
 }
 
 impl WasmState {
```

Comment:
FYI my intention behind this WasmState struct was to implement ModuleImportResolver and Externals directly on it.

Reply:
I did think of it, but when creating a `wasmi::ModuleInstance` we have to pass in a resolver). Since this is happening in the `WasmState::new` function, there is no `self` to pass as the resolver, if `ModuleImportResolver` and `Externals` were to be implemented for `WasmState`.

Reply:
Would something like this work? (I guess probably not, since the references would overlap, but maybe there is a way to make it work).

```rust
let wasm_state = WasmState { ... };
let instance = ModuleInstance::new(module, ...&wasm_state);
let wasm_state.instance = instance;
Ok(wasm_state)
```

If not anyways, then I think we can probably remove the `WasmState` wrapper, since (in my mind) that was its only purpose.